opensearch-project/anomaly-detection
# Adding Model Type Validation to Validate API ("non-blocker")

### Description
This PR adds two new changes.

1.  Adds model (“non-blocker”) type validation to the current validation API. The model type validation is non-blocking in nature meaning that any warning that arise are simply suggestions or added information to the user about which configuration could be changed in order to increase the likelihood of model training completing successfully. These validation checks could also potentially recommend a specific detector interval or window delay value that the user can try to use.
    * Every-time model type validation is called, it initially also calls the blocker validation since there its not possible to run non-blocker checks if there is a blocking issue initially. 
2. Adds a check that the given timefield exists in the index mapping and also that it is of type date to AbstractAnomalyDetectorActionHandler which means this check is now executed for model

### Non-Blocker validation steps:

1. Get latest time stamp (if no latest time stamp found then user is told there is not enough historical data)
2. Check if detector is of type multi-entity and if it is then only look at the top entity for any proceeding validation
3. Check if general density of data is good enough for current configurations to complete model training with all the configurations applied.
    * This check will be run multiple times with different interval configurations until an interval is either recommended, same as the given one or no interval can be found
4. If no interval can be found with all configurations applied that will lead to data that is dense enough then proceed to run general density check by sequentially adding configurations one at a time
    * For example, first only the raw data is looked at in terms of how dense the data is →  if raw data on its own is dense enough then next the data filter (filter_query) will be added and general density will be checked for again
5. If no other issues exist in the data then check if any new window delay recommendation is possible (latest time stamp is used)

### Additional Info (Important):

* All validation for general density involves a bucket aggregation that finds if enough buckets in the last x intervals have at least 1 document present.
* I will be adding more new tests in separate commit or PR
* Response wording haven't finished tech writer review yet
### API Request:
* Path: `_plugins/_anomaly_detection/detectors/_validate/<aspect>`
* Parameter(aspect)
    * `model`
* Payload: Detector Configs
 
### Examples
 #### Example 1 (data ingested only every 5 mins, recommends 5 minutes interval of 1 minute) :
 ```
{
    "name": "test-hc-detector",
    "description": "Test detector",
    "time_field": "@timestamp",
    "indices": [
        "host-cloudwatch-two"
    ],
    "feature_attributes": [
        {
            "feature_name": "test",
            "feature_enabled": true,
            "aggregation_query": {
                "test": {
                    "avg": {
                        "field": "cpu"
                    }
                }
            }
        }
    ],
    "filter_query": {
        "bool": {
            "filter": [
                {
                    "exists": {
                        "field": "host"
                    }
                }
            ],
            "adjust_pure_negative": true,
            "boost": 1
        }
    },
    "detection_interval": {
        "period": {
            "interval": 1,
            "unit": "Minutes"
        }
    },
    "window_delay": {
        "period": {
            "interval": 1,
            "unit": "Minutes"
        }
    },
    "category_field": [
        "host"
    ]
}
```

#### Response 1:

```
{
    "model": {
        "detection_interval": {
            "message": "We suggest using a detector interval of: 5",
            "suggested_value": {
                "period": {
                    "interval": 5,
                    "unit": "Minutes"
                }
            }
        }
    }
}
```
#### Example 2 request (filter query leads to not enough dense data)
* CPU values in this data set range between 5-15 minutes
```
{
    "name": "test-hc-detector",
  ...
   "filter_query": {
        "bool": {
            "filter": [
                {
                    "range": {
                        "cpu": {
                            "gt": 10000
                        }
                    }
                }
            ],
            "adjust_pure_negative": true,
            "boost": 1
        }
    }
    "detection_interval": {
        "period": {
            "interval": 5,
            "unit": "Minutes"
        }
    },
    ...
}
```
#### Response 2: 
```
{
    "model": {
        "filter_query": {
            "message": "Data is too sparse after data filter is applied."
        }
    }
}
```

### Check List

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
